### PR TITLE
Feature/2309/workflow entryname hosted and apps

### PR DESCRIPTION
**Pull request by Richard-Hansen**
Ticket: https://ucsc-cgl.atlassian.net/browse/DOCK-633

Fixing previous issues regarding handling HOSTED tools and workflows and DOCKSTORE_YML workflows. 
Added integration tests.

Changes were very small, hiding whitespace will help.

### CODE REVIEW

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java [hunk 1]
```diff
@@ -704,25 +705,30 @@ protected void handlePublishUnpublish(String entryPath, String newName, boolean
...
-
-                    out("Successfully registered " + completeEntryPath);
-
-                    workflowsApi.refresh(newWorkflow.getId(), true);
-                    publish(true, completeEntryPath);
-                } catch (ApiException ex) {
-                    exceptionMessage(ex, "Unable to publish " + entryPath + "/" + newName, Client.API_ERROR);
+                // Prevent specifying a custom name for both HOSTED and DOCKSTORE_YML workflows
+                if (Workflow.ModeEnum.HOSTED.equals(existingWorkflow.getMode()) || Workflow.ModeEnum.DOCKSTORE_YML.equals(existingWorkflow.getMode())) {
+                    out(WorkflowClient.INVALID_WORKFLOW_MODE_PUBLISH);
```
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:710]: Should be `err` not `out`? Or even `errorMessage`? Don't we want the program to exit with a non-zero exit code if it fails?

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java [hunk 2]
```diff
@@ -704,25 +705,30 @@ protected void handlePublishUnpublish(String entryPath, String newName, boolean
...
+                        );
+
+                        final String completeEntryPath = entryPath + "/" + newName;
+
+                        out("Successfully registered " + completeEntryPath);
+
+                        workflowsApi.refresh(newWorkflow.getId(), true);
+                        publish(true, completeEntryPath);
+                    } catch (ApiException ex) {
+                        exceptionMessage(ex, "Unable to publish " + entryPath + "/" + newName, Client.API_ERROR);
```
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:730]: I think you can reuse `completeEntryPath`

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java [hunk 3]
```diff
@@ -704,25 +705,30 @@ protected void handlePublishUnpublish(String entryPath, String newName, boolean
...
-                    final String completeEntryPath = entryPath + "/" + newName;
-
-                    out("Successfully registered " + completeEntryPath);
-
-                    workflowsApi.refresh(newWorkflow.getId(), true);
-                    publish(true, completeEntryPath);
-                } catch (ApiException ex) {
-                    exceptionMessage(ex, "Unable to publish " + entryPath + "/" + newName, Client.API_ERROR);
+                // Prevent specifying a custom name for both HOSTED and DOCKSTORE_YML workflows
+                if (Workflow.ModeEnum.HOSTED.equals(existingWorkflow.getMode()) || Workflow.ModeEnum.DOCKSTORE_YML.equals(existingWorkflow.getMode())) {
```
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:709]: You can and should compare enums with `==`

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/ToolClient.java [hunk 4]
```diff
@@ -238,32 +239,36 @@ protected void handlePublishUnpublish(String entryPath, String newName, boolean
...
-
-                    newContainer = containersApi.registerManual(newContainer);
-
-                    out("Successfully registered " + entryPath + "/" + newName);
-
-                    containersApi.refresh(newContainer.getId());
-                    publish(true, newContainer.getToolPath());
-                } catch (ApiException ex) {
-                    exceptionMessage(ex, "Unable to publish " + newName, Client.API_ERROR);
+                if (DockstoreTool.ModeEnum.HOSTED.equals(existingTool.getMode())) {
```
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/ToolClient.java:242]: Compare with `==`

**Comment by Richard-Hansen** [comment]
Looks like there are some unexpected TravisCI errors, I'll remove reviewers until after fixing them.

Sorry for the email clutter.

**Comment by denis-yuen** [comment]
FYI, this was mostly about trying out the various changes to the build, webservice (cache, rate limit behaviour)

**Review by coverbeck — APPROVED** [review]

**Review by svonworl — APPROVED** [review]

**Pull request merged**